src/App.js

```diff
@@ -1,314 +1,57 @@
-//   };
-
-//   const checkIfTie = () => {
-//     let filled = true;
-//     board.forEach((square) => {
-//       if (square == "") {
-//         filled = false;
-//       }
-//     });
-
-//     if (filled) {
-//       setResult({ winner: "No One", state: "Tie" });
-//     }
-//   };
-
-//   const restartGame = () => {
-//     setBoard(["", "", "", "", "", "", "", "", ""]);
-//     setPlayer("O");
-//   };
-
-  
-//     // Use two loops to make the squares
-//     const boardSize = 3;
-//     let squares = [];
-//     for (let i = 0; i < boardSize; ++i) {
-//       let row = [];
-//       for (let j = 0; j < boardSize; ++j) {
-//         row.push(
-//           <Square
-//             val={board[i * boardSize + j]}
-//             chooseSquare={() => {
-//               chooseSquare(i * boardSize + j);
-//             }}
-//           />
-//         );
-//        // squares.push(<div className="row"></div>); remove key;
-//               squares.push(
-//                 <div key={i} className="row"> 
-//                   {row}
-//                 </div>
-//               );
-
-//       }
-//       console.log(squares);
-//     }
-//     return (
-//       <div className="App">
-//         <div className="board">
-//           <div>{squares}</div>
-//         </div>
-//       </div>
-//     );
-  
-//   }
-// export default App;
-
 import "./App.css";
 import { useState, useEffect } from "react";
 import Square from "./Components/Square";
 import { Patterns } from "./Patterns";
 
 function App() {
   const [board, setBoard] = useState(["", "", "", "", "", "", "", "", ""]);
-  const [player, setPlayer] = useState("O");
-  const [result, setResult] = useState({ winner: "none", state: "none" });
-
+  // const [player, setPlayer] = useState("O");
+  // const [result, setResult] = useState({ winner: "none", state: "none" });
+  const O_TEXT = "0";
+  const X_TEXT = "X";
   useEffect(() => {
     checkWin();
     checkIfTie();
 
-    if (player == "X") {
-      setPlayer("O");
-    } else {
-      setPlayer("X");
-    }
+    // if (player == "X") {
+    //   setPlayer("O");
+    // } else {
+    //   setPlayer("X");
+    // }
   }, [board]);
 
   useEffect(() => {
-    if (result.state != "none") {
-      alert(`Game Finished! Winning Player: ${result.winner}`);
+    if (checkWin().foundWinningPattern == true) {
+      const currentPlayer = currentPlayerFromState();
+      alert(`Game Finished! Winning Player: ${currentPlayer}`);
+      restartGame();
+    } else if (checkIfTie.filled == "true") {
+      alert(`Game Finished! Winning Player: none`);
       restartGame();
     }
-  }, [result]);
+  });
 
   const chooseSquare = (square) => {
     setBoard(
       board.map((val, idx) => {
         if (idx == square && val == "") {
-          return player;
+          const currentPlayer = currentPlayerFromState();
+          return currentPlayer;
         }
 
         return val;
       })
     );
   };
 
+  const currentPlayerFromState = () => {
```

Comment:
I'd recommend passing in `board` into the function instead of using a closure here. Not only would passing it in be cleaner but also it would let you write unit tests for it.

Reply:
I'd probably expect unit tests for this sort of function BTW. Policies vary from company to company on this sort of thing though.

Reply:
I will get started writing a unit test for this function. I am also thinking about writing a unit test for 'chooseSquare'.

Reply:
I get an error 'TypeError: Cannot read properties of undefined (reading 'filter')'
Am I getting the error because the filter does not execute the function for empty elements?
In the previous codebase, 'm sure the filter function was used in a similar way.

Reply:
When you get an error message like this I'd type the generic portions of the message into google and see what comes up. Most common errors will have some useful blog posts about it. The filter should go through all elements of the array, even any undefined values. Instead, I think the value of "board" is undefined. See https://rollbar.com/blog/javascript-typeerror-cannot-read-property-of-undefined/